src/Sarif/ZipArchiveArtifact.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+
+    public class ZipArchiveArtifact : IEnumeratedArtifact
+    {
+        private readonly ISet<string> binaryExtensions;
+        private readonly ZipArchive archive;
+        private ZipArchiveEntry entry;
+        private readonly Uri uri;
+        private string contents;
+        private byte[] bytes;
+
+        public ZipArchiveArtifact(ZipArchive archive, ZipArchiveEntry entry, ISet<string> binaryExtensions)
+        {
+            this.entry = entry;
+            this.archive = archive;
+            this.binaryExtensions = binaryExtensions;
+            this.uri = new Uri(entry.FullName, UriKind.RelativeOrAbsolute);
+        }
+
+        public Uri Uri => this.uri;
+
+        public Stream Stream
+        {
+            get
+            {
+                lock (this.archive)
+                {
+                    return entry != null
+                        ? entry.Open()
+                        : null;
+                }
+            }
+            set => throw new NotImplementedException();
+        }
+
+        public Encoding Encoding { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
+
+        public string Contents
+        {
+            get => GetArtifactData().text;
+            set => this.contents = value;
+        }
+
+        public byte[] Bytes
+        {
+            get => GetArtifactData().bytes;
+            set => this.bytes = value;
+        }
+
+        private (string text, byte[] bytes) GetArtifactData()
+        {
+            if (this.contents == null && this.bytes == null)
+            {
+                lock (this.archive)
+                {
+                    if (this.contents == null && this.bytes == null)
+                    {
+                        string extension = Path.GetExtension(Uri.ToString());
+                        if (this.binaryExtensions.Contains(extension))
+                        {
+                            this.bytes = new byte[Stream.Length];
```

Comment:
Do all our underlying Streams support .Length?

It is more defensive to promote these .bytes fields to MemoryStreams and use Stream.Copy.  If I'm reading the docs right: wherever the underlying Stream supports .CanSeek, the MemoryStream can be precisely presized by safely querying .Length.  Otherwise, .Length can throw for some implementations (but not others.)

The tradeoff of doing this is that if we ever needed a precisely-sized byte[] for an artifact later, we would potentially need to take a copy.  (But not in cases where we were able to size it precisely.)

Reply:
ZipArchive supports providing a decompressed length. What's the performance impact of creating a memory stream as you describe? Will it effectively use the wrapped stream as its buffer? And so the length computation is computable before the memory stream did it? :)

This sounds like a nice next clean-up. It also makes me wonder if I've been over-concerned about handling non-seekable streams. This case may be entirely unremarkable based on what you've pointed out: we should simply read the original stream into a byte array, sniff it and then, if we need to decode, we wrap it in a memory stream (which will use the byte array as its buffer and not create a new copy).

Reply:
OK. But for this specific location, ZipArchive, these streams are never seekable. Do you have a have suggestion or do you agree this code can ship as is? :)